### PR TITLE
Scijava/scijava ops engine/skip map history

### LINKED ISSUE #99: TamuraFeature2dTest#testCoarsenessFeature AIOOBE
Following from https://github.com/scijava/incubator/pull/62#issuecomment-1282811260

When running [testCoarsenessFeature](https://github.com/scijava/incubator/blob/eaa8d436c30bed584c0be6184f82f778ac36adda/imagej/imagej-ops2/src/test/java/net/imagej/ops2/features/tamura2d/Tamura2dFeatureTest.java#L64-L65) in an IDE (IntelliJ, windows) it will throw an `ArrayIndexOutOfBoundsException` with the following stack trace:

```
Caused by: java.lang.ArrayIndexOutOfBoundsException: Index 10000 out of bounds for length 10000
	at net.imglib2@5.12.0/net.imglib2.img.basictypeaccess.array.AbstractByteArray.getValue(AbstractByteArray.java:61)
	at net.imglib2@5.12.0/net.imglib2.type.numeric.integer.GenericByteType.getByte(GenericByteType.java:137)
	at net.imglib2@5.12.0/net.imglib2.type.numeric.integer.GenericByteType.valueEquals(GenericByteType.java:243)
	at net.imglib2@5.12.0/net.imglib2.type.numeric.integer.GenericByteType.equals(GenericByteType.java:253)
	at java.base/java.util.WeakHashMap.eq(WeakHashMap.java:287)
	at java.base/java.util.WeakHashMap.getEntry(WeakHashMap.java:431)
	at java.base/java.util.WeakHashMap.containsKey(WeakHashMap.java:418)
	at org.scijava.ops.engine@0-SNAPSHOT/org.scijava.ops.engine.DefaultOpHistory.logOutput(DefaultOpHistory.java:77)
	at org.scijava.ops.engine@0-SNAPSHOT/org.scijava.ops.engine.matcher.impl.AbstractRichOp.postprocess(AbstractRichOp.java:54)
	at org.scijava.ops.engine@0-SNAPSHOT/org.scijava.ops.engine.matcher.impl.OpWrappers$Computer1OpWrapper$1GenericTypedComputer1.compute(OpWrappers.java:806)
	at net.imagej.ops2@0-SNAPSHOT/net.imagej.ops2.map.neighborhood.MapNeighborhoodAllRAI.lambda$compute$0(DefaultMapNeighborhood.java:116)
	at net.imglib2.loops.BindActionToSamplers$BiConsumerRunnable.run(BindActionToSamplers.java:144)
	at net.imglib2.loops.LoopUtils$LineProcessor.run(LoopUtils.java:147)
	at net.imglib2.loops.LoopUtils$LineProcessor.run(LoopUtils.java:147)
	at net.imglib2@5.12.0/net.imglib2.loops.LoopBuilder.lambda$runOnChunkUsingRandomAccesses$3(LoopBuilder.java:367)
	at net.imglib2@5.12.0/net.imglib2.loops.LoopBuilder.lambda$forEachPixel$0(LoopBuilder.java:159)
	at net.imglib2@5.12.0/net.imglib2.loops.LoopBuilder.runOnChunkUsingRandomAccesses(LoopBuilder.java:365)
	at net.imglib2@5.12.0/net.imglib2.loops.LoopBuilder.lambda$runUsingRandomAccesses$2(LoopBuilder.java:356)
	at net.imglib2@5.12.0/net.imglib2.parallel.DefaultTaskExecutor.lambda$forEachApply$1(DefaultTaskExecutor.java:118)
	at java.base/java.util.concurrent.ForkJoinTask$AdaptedCallable.exec(ForkJoinTask.java:1448)
	at java.base/java.util.concurrent.ForkJoinTask.doExec(ForkJoinTask.java:290)
	at java.base/java.util.concurrent.ForkJoinPool$WorkQueue.topLevelExec(ForkJoinPool.java:1020)
	at java.base/java.util.concurrent.ForkJoinPool.scan(ForkJoinPool.java:1656)
	at java.base/java.util.concurrent.ForkJoinPool.runWorker(ForkJoinPool.java:1594)
	at java.base/java.util.concurrent.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:177)
```

Some observations:
* If I comment out [this line](https://github.com/scijava/incubator/blob/eaa8d436c30bed584c0be6184f82f778ac36adda/scijava/scijava-ops-engine/src/main/java/org/scijava/ops/engine/DefaultOpHistory.java#L77) in the `DefaultOpHistory`, I do not see the exception
* If I [make this LoopBuilder](https://github.com/scijava/incubator/blob/eaa8d436c30bed584c0be6184f82f778ac36adda/imagej/imagej-ops2/src/main/java/net/imagej/ops2/map/neighborhood/DefaultMapNeighborhood.java#L114) **not multithreaded**, I do not see this exception

This leads me to believe that the `OpHistory` is not threadsafe.

### PR DESCRIPTION
**Pull request by hinerm**
Closes https://github.com/scijava/scijava/issues/99

See also [#100](https://github.com/scijava/scijava/issues/100) and [#101](https://github.com/scijava/scijava/issues/101)

### CODE REVIEW

File: imagej/imagej-ops2/src/main/java/net/imagej/ops2/map/neighborhood/MapNeighborhoodWithCenter.java [hunk 1]
```diff
@@ -118,10 +120,20 @@ public void compute(final RandomAccessibleInterval<I> in1, final Shape in2,
...
+		}
+
 		LoopBuilder.setImages(neighborhoodInput, in1, out).multiThreaded()
 			.forEachPixel((neighborhood, inPixel, outPixel) -> {
 				centerAwareOp.compute(neighborhood, inPixel, outPixel);
 			});
+
+		if (centerAwareOp instanceof RichOp) {
+			((RichOp)centerAwareOp).metadata().hints().minus(
+					BaseOpHints.History.SKIP_RECORDING);
```
Comment (ctrueden) [inline imagej/imagej-ops2/src/main/java/net/imagej/ops2/map/neighborhood/MapNeighborhoodWithCenter.java:141]: If `SKIP_RECORDING` hint was already set before getting plussed in above... then should we still be minusing it out again here? Don't we just want to restore the hints back to their original state?

**Review by ctrueden — APPROVED** [review]
Looks awesome. I commented on one small thing, but I think we should merge this as-is as soon as possible, to get the nondeterministic failures under control.

**Comment by ctrueden** [comment]
On this branch, I do get one failure in imagej-ops2, on both the CLI and in IntelliJ:
```
[ERROR] Tests run: 3, Failures: 0, Errors: 1, Skipped: 0, Time elapsed: 0.755 s <<< FAILURE! - in net.imagej.ops2.features.tamura2d.Tamura2dFeatureTest
[ERROR] testCoarsenessFeature  Time elapsed: 0.639 s  <<< ERROR!
java.lang.ArrayIndexOutOfBoundsException
	at net.imagej.ops2@0-SNAPSHOT/net.imagej.ops2.features.tamura2d.Tamura2dFeatureTest.testCoarsenessFeature(Tamura2dFeatureTest.java:65)
Caused by: java.lang.ArrayIndexOutOfBoundsException: Index 10000 out of bounds for length 10000
```

According to the stack trace, it's still trying to log the output of individual pixels. It's like the hint isn't being respected.

**Review by ctrueden — CHANGES_REQUESTED** [review]
On second thought, I think we better get to the bottom of the hint-propagation failure before merging this. I strongly considered just commenting out the offending test, but it's going to bite us sooner rather than later if we do that.

**Comment by hinerm** [comment]
> On second thought, I think we better get to the bottom of the hint-propagation failure before merging this. I strongly considered just commenting out the offending test, but it's going to bite us sooner rather than later if we do that.

@ctrueden see https://github.com/scijava/scijava/issues/101#issuecomment-1289343895

**Comment by hinerm** [comment]
@ctrueden @gselzer I just added a commit allowing the updating of `Hints` in an `OpMetadata`.. these tests are passing consistently for me now. Can you test a few times on your end and see if you get any failures?

**Review by ctrueden — APPROVED** [review]
<details><summary>With OpenJDK 17.0.3, there is an unrelated JPMS issue with scijava-persist</summary>

```
[INFO] Running org.scijava.persist.SerializationTests
IClassAdapters : 
IRunTimeClassAdapters : 
	 interface org.scijava.persist.Shape
	 	 class org.scijava.persist.Circle
	 	 class org.scijava.persist.Shapes$Diamond
	 	 class org.scijava.persist.Shapes$Rectangle
IClassAdapters : 
IRunTimeClassAdapters : 
	 interface org.scijava.persist.Shape
	 	 class org.scijava.persist.Circle
	 	 class org.scijava.persist.Shapes$Diamond
	 	 class org.scijava.persist.Shapes$Rectangle
[ERROR] Tests run: 2, Failures: 0, Errors: 2, Skipped: 0, Time elapsed: 0.213 s <<< FAILURE! - in org.scijava.persist.SerializationTests
[ERROR] testCircleObject  Time elapsed: 0.192 s  <<< ERROR!
java.lang.reflect.InaccessibleObjectException: Unable to make field int org.scijava.persist.Circle.radius accessible: module org.scijava.persist does not "opens org.scijava.persist" to module com.google.gson
	at org.scijava.persist@0-SNAPSHOT/org.scijava.persist.SerializationTests.testSerializationDeserialization(SerializationTests.java:76)
	at org.scijava.persist@0-SNAPSHOT/org.scijava.persist.SerializationTests.testCircleObject(SerializationTests.java:50)

[ERROR] testDrawingObject  Time elapsed: 0.012 s  <<< ERROR!
java.lang.reflect.InaccessibleObjectException: Unable to make field int org.scijava.persist.Circle.radius accessible: module org.scijava.persist does not "opens org.scijava.persist" to module com.google.gson
	at org.scijava.persist@0-SNAPSHOT/org.scijava.persist.SerializationTests.testSerializationDeserialization(SerializationTests.java:76)
	at org.scijava.persist@0-SNAPSHOT/org.scijava.persist.SerializationTests.testDrawingObject(SerializationTests.java:62)
```

</details>

With OpenJDK 11.0.16, I got the following results:

1. BUILD SUCCESS on Ubuntu
2. BUILD SUCCESS on macOS
3. BUILD FAILURE on Ubuntu: `org.scijava.ops.api.features.OpMatchingException: No MatchingRoutine was able to produce a match! at net.imagej.ops2@0-SNAPSHOT/net.imagej.ops2.identity.IdentityTest.testIdentity(IdentityTest.java:49)`
4. BUILD SUCCESS on macOS
5. BUILD SUCCESS on Ubuntu

This `No MatchingRoutine was able to produce a match!` error is not new; it was happening intermittently previously, and will need investigation, but I do not think it should hold up merging of this PR, which does seem to address the AIOOBEs and likely performance issues of map operations.

**Pull request merged**